src/ScottPlot/Plot.cs

```diff
@@ -225,7 +227,10 @@ public void SaveFig(string filePath, bool renderFirst = true)
 
         public void Add(Plottable plottable)
         {
-            settings.plottables.Add(plottable);
+            if (DiagnosticMode)
+                settings.plottables.Add(new DecoratorFactory().WrapWithAll(plottable));
```

Comment:
This *is* a breaking change even though `PlottableDecoratorBase` extends `Plottable`

For example, if the user has code like this:
```cs
if(plottables[0].GetType() == typeof(PlottableOHLC)){
    //Foo
}
```

This is why C# has the `is` operator, which works correctly for childclasses:

```cs
if(plottables[0] is PlottableOHLC){
    //Foo
}
```

Now, ScottPlot doesn't make many guarantees about backwards compatibility when it comes to the actual plottables, because users are expected to use `Plot.PlotOHLC()` instead of using the constructors, and those methods are guaranteed to be backwards compatible. Furthermore, I don't think `settings.plottables` makes any backwards compatibility guarantees because users aren't expected to use the `Settings` object at all.

I understand this concern is pretty minor, and it's unclear whether this is supposed to be guaranteed backwards compatible code. Especially given that these issues will only exist in diagnostic mode which is for development only. And if these changes are too breaking, it could be merged into ScottPlot 5.0 instead of 4.0.40.

Reply:
Hey @Benny121221, this is a good point to be aware of, and your message is well-received. Breaking changes are something to keep an eye-out for, and _run time_ breaking changes could be especially difficult to troubleshoot for users not expecting it.

A discussion could be had about whether this is or is not the best architecture to use for data verification (e.g., adding a simple `bool verifyBeforeRender` and `VerifyNow()` to the `Plottable` class or maybe to a new `IVerifiable` interface may be preferred to this inheritance system), but maintaining strict backward-compatibility is not one of my chief concerns.

> ScottPlot doesn't make many guarantees about backwards compatibility

> it's unclear whether this is supposed to be guaranteed backwards compatible code

Backwards compatibility is something we prefer, but it's not a hardened rule. ScottPlot has a bit of a history of sneaking small breaking changes into the code base when a breaking change confers a real benefit to reducing bugs, improving performance, adding new useful features, or improving maintainability of the code base. I think the number of users is still small enough that I don't mind making small breaking changes which may only affect fringe use cases. 

The expectation of possible slight breaking changes is one reason [every version of ScottPlot has its own cookbook](https://swharden.com/scottplot/cookbook).

All that said, I think I'm going to release ScottPlot 4.0.40, merge this into ScottPlot 4.1.0, and make a couple larger breaking changes at the same time.

Reply:
> This is a breaking change even though PlottableDecoratorBase extends Plottable

Not really, it wraps around a real object. So you can't find it with `is` too. `Decorator` contain source plottable in protected `sourcePlottable` field. 
I proceeded from the assumption that the `settings.plottables` list is not available to the user and no one manipulates it manually in their projects. But you gave me the idea that he was additionally manipulated by Clear() methods and they check for type like you provided. I think I'll need to tweak them too.
>Now, ScottPlot doesn't make many guarantees ...

The key here is that when the diagnostic mode is turned off, ScottPlot works exactly the same as before. `settings.plottables` contain real plottable classes like `PlottableOHLC`. And none of these decorators are used at all. Only one check of the DiagnosticMode flag when adding Plottable. 

Regardless of the diagnostic mode, the user can add Plottable through the `Plot.Plot....` as before. api, or create manually using constructors and Add it with 'Plot.Add(new PlottableOHLC())'.

The diagnostic mode itself is designed for single use only.
When the user sees an incomprehensible library behavior, he turns `DiagnosticMode`, checks his data with it, fix errors in the data if any, and turns it off. If someone decides to enable it by default, then I am strongly opposed, this is a bunch of costly, unnecessary checks also through reflection, I do not approve of this.

> And if these changes are too breaking, it could be merged into ScottPlot 5.0 instead of 4.0.40.

Yes, this is not an urgent fix, and can be postponed for a long time. All the more, on good terms, it should be expanded to dozens of checks and go through all the existing Plotables.

I'm sorry if I didn't understand you (my poor English), but you didn't convince me about breaking changes.

Reply:
`PlottableDecoratorBase` inherits from `Plottable`, which means the `is` operator will return `true` but `typeof`/`GetType()` will not. An example is available [here](https://tio.run/##dY49CsJAEIVr9xRTbgpzAEMKEbRREBQsxGKJmzCQ7ISdiRJCzr6uSSzt3s/34BW8LsjbEDpGV8GlZ7FNporaMMPZU@VNA4NasRjBAl6ETzgZdJrFx8H9AcZXnHyR1bZtawslEeTg7Bsmr5MsVjtyTLVNbx7FHtFZHbH0YOXatxGBPAeJikq99x1K8n8EyDAzERnVqJavUzaMi5uvbH5pCB8)

In my opinion, using the `is` operator is probably best practice anyways, especially when it comes to third-party libraries that may not guarantee types but only type compatibility. Plus the `is` operator is more convenient. I have seen style guides forbid `typeof`/`GetType()` except in situations where you cannot accept subclasses (which should never exist as it violates polymorphism). However, Microsoft doesn't give any warning against its use (that I can find).

As for situations where this matters, I can't think of very many. The best example I have is this code which would break if it used `typeof`/`GetType()` in diagnostic mode: https://github.com/Benny121221/WPlot/blob/master/MainWindow.xaml.cs#L388L399

Given that it only happens in diagnostic mode with questionable code, I don't think it matters.

Reply:
@Benny121221 , I understood what you meant, and you are absolutely right (I missed existing `GetPlottables()`), 
If the user can do something using the existing api, it is wrong to impose additional implicit restrictions.